src/bussinessLogic/ImageManager.java

```diff
@@ -15,7 +15,7 @@
 	private static final int numberOfFiles = 14;
 	private static int[] imageList;
 
-	public ImageManager(int[] imageList) {
+	public void setImageList (int[] imageList) {
```

Comment:
De ce ai inlocuit constructorul cu aceasta metoda?

Reply:
Cum ziceam, ca sa fie codul mai usor de mentinut aveam nevoie ca _images_ sa fie calculat in functie de numarul de coloane introduse la initializarea obiectului StartUp. Constructorul ImageManager preia array-ul images in momentul instantierii obiectului, dar in acelasi timp obiectul StartUp are nevoie de obiectul ImageManager la instantiere, deci nu-i poate inca furniza array-ul (null pointer exception). Asa ca m-am gandit  ca e mai usor sa creez o metoda care sa fie chemata la nevoie decat sa ma joc de-a "Inception" cu cele doua obiecte :smile: Este vreo solutie mai "eleganta" la aceasta problema sau e ok asa?

Reply:
Solutia eleganta este aceea de a initializa in constructor aceasta variabila. Mai apoi, valoarea ei sa fie folosita in restul codului. Daca undeva apare _null pointer exception_ atunci inseamna ca altundeva este o problema in logica sau fluxul programului.

Reply:
_Null pointer exception_ primesc in momentul in care pun _images_ ca parametru la instantierea obiectului ImageManager (pentru ca images nu e inca initializat). Daca pun cum am pus inainte (_new int_[12]) merge, doar ca asta nu e o practica buna: de ex. daca pe viitor voi fi nevoit sa modific numarul coloanelor sau randurilor (sa fac 8x6 de exemplu) va trebui sa fiu atent sa modific parametrii si la instantierea obiectului StartUp si la ImageManager. Si daca mai adaug si alte metode in ImageManager care vor avea nevoie de alti parametrii care sunt initializati in metoda initialize() (numberOfColumns de exemplu) atunci chiar am dat-o in bara, fiindca va trebui sa modific manual fiecare parametru pe care-l primeste ImageManager ca sa corespunda cu parametrii pe care-i calculeaza metoda _initialize()_, deci sansele de a introduce o eroare cresc substantial. Propriu-zis in momentul de fata nu pot initializa ImageManager daca nu e initializat StartUp, dar in acelasi timp nu pot initializa StartUp fara ImageManager :confused:

Reply:
> daca pe viitor voi fi nevoit sa modific numarul coloanelor sau randurilor (sa fac 8x6 de exemplu) va trebui sa fiu atent sa modific parametrii si la instantierea obiectului StartUp si la ImageManager.

Cele doua entitati trebuie sa fie versatile. Daca exista o dependenta ca si cea descrisa de tine mai sus, atunci aceasta trebuie adresata.


> Si daca mai adaug si alte metode in ImageManager care vor avea nevoie de alti parametrii care sunt initializati in metoda initialize() (numberOfColumns de exemplu) atunci chiar am dat-o in bara, fiindca va trebui sa modific manual fiecare parametru pe care-l primeste ImageManager ca sa corespunda cu parametrii pe care-i calculeaza metoda initialize()

Din descrierea ta, imi dau seama ca entitatea _ImageManager_ nu este decuplata de restul programului. In mod normal asa ar trebui sa fie. Decuplat, inseamna ca trebuie sa aibe o oarecare independenta.

Daca va fi nevoie de alte metode si alti parametrii, atunci se prea poate sa fie nevoie sa se creeze o clasa noua care sa adreseza noua functionalitate.

Reply:
> Cele doua entitati trebuie sa fie versatile. Daca exista o dependenta ca si cea descrisa de tine mai sus, atunci aceasta trebuie adresata.

La ce te gandesti cand spui ca trebuie adresata? (adica in ce mod?)

> Din descrierea ta, imi dau seama ca entitatea _ImageManager_ nu este decuplata de restul programului. In mod normal asa ar trebui sa fie. Decuplat, inseamna ca trebuie sa aibe o oarecare independenta.

Pai da; si aceasta dependenta e cauzata tocmai de parametrul _images_ care trebuie trimis constructorului, el nefiind inca initializat (sau, ma rog, e initializat la valoare nula).

Reply:
> > Cele doua entitati trebuie sa fie versatile. Daca exista o dependenta ca si cea descrisa de tine mai sus, atunci aceasta trebuie adresata.
> 
> La ce te gandesti cand spui ca trebuie adresata? (adica in ce mod?)  -> **Codul trebuie scris in asa fel incat aceasta dependenta sa nu existe.**
> 
> > Din descrierea ta, imi dau seama ca entitatea _ImageManager_ nu este decuplata de restul programului. In mod normal asa ar trebui sa fie. Decuplat, inseamna ca trebuie sa aibe o oarecare independenta.
> 
> Pai da; si aceasta dependenta e cauzata tocmai de parametrul _images_ care trebuie trimis constructorului, el nefiind inca initializat (sau, ma rog, e initializat la valoare nula). -> **O sa ma uit peste cod sa vad ce modificari pot face in acest sens, iar mai apoi o sa fac pull request cu cateva modificari.**

Raspunsurile sunt cu litere ingrosate.